lib/fiddle/struct.rb

```diff
@@ -94,33 +109,53 @@ def create(klass, types, members)
           define_method(name){ @entity[name] }
           define_method(name + "="){|val| @entity[name] = val }
         }
+        define_singleton_method(:malloc) do |func=nil|
+          struct = new(entity_class.malloc(types, func, size))
+          if block_given?
+            raise ArgumentError, 'a free function must be supplied to malloc when it is called with a block' unless func
+            begin
+              yield struct
+            ensure
+              struct.to_ptr.call_free
+            end
+          else
+            struct
+          end
+        end
       }
       size = klass.entity_class.size(types)
+      # we use eval here make size into a literal, for performance
       new_class.module_eval(<<-EOS, __FILE__, __LINE__+1)
         def new_class.size()
           #{size}
         end
-        def new_class.malloc()
-          addr = Fiddle.malloc(#{size})
-          new(addr)
-        end
       EOS
       return new_class
     end
     module_function :create
   end
 
-  # A C struct wrapper
+  # A pointer to a C structure
   class CStructEntity < Fiddle::Pointer
     include PackInfo
     include ValueUtil
 
     # Allocates a C struct with the +types+ provided.
     #
     # See Fiddle::Pointer.malloc for memory management issues.
-    def CStructEntity.malloc(types, func = nil)
-      addr = Fiddle.malloc(CStructEntity.size(types))
-      CStructEntity.new(addr, types, func)
+    def CStructEntity.malloc(types, func = nil, size = size(types), &block)
+      struct = super(size, func, &nil)
+      struct.set_ctypes types
+      if block_given?
+        raise ArgumentError, 'a free function must be supplied to Fiddle::CStructEntity.malloc when it is called with a block' unless func
+        begin
+          yield struct
+        ensure
+          struct.call_free
+        end
+      else
+        struct
+      end
```

Comment:
How about reusing `malloc` block implementation?

If we reuse `malloc` block implementation, the allocated pointer is freed immediately when an exception is raised in `set_ctypes`.

```suggestion
      if block_given?
        super(size, func) do |struct|
          struct.set_ctypes types
          yield struct
        end
      else
        struct = super(size, func)
        struct.set_ctypes types
        struct
      end
```